bitcoin/src/blockdata/transaction/extended.rs

```diff
@@ -0,0 +1,133 @@
+/// exist.
+#[derive(Clone, Debug)]
+pub struct ExtendedTransaction<ExtraTxInData = ()> {
+    /// The protocol version, is currently expected to be 1 or 2 (BIP 68).
+    pub version: Version,
+
+    /// List of extended transaction inputs.
+    pub inputs: Vec<ExtendedTxIn<ExtraTxInData>>,
+
+    /// List of transaction outputs.
+    pub outputs: Vec<TxOut>,
+
+    /// Block height or timestamp. Transaction cannot be included in a block until this height/time.
+    ///
+    /// ### Relevant BIPs
+    ///
+    /// * [BIP-65 OP_CHECKLOCKTIMEVERIFY](https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki)
+    /// * [BIP-113 Median time-past as endpoint for lock-time calculations](https://github.com/bitcoin/bips/blob/master/bip-0113.mediawiki)
+    pub lock_time: absolute::LockTime,
+}
+
+impl<T> ExtendedTransaction<T> {
+    /// Converts extended transaction into non-extended transaction.
+    ///
+    /// This performs one allocation for the vec of inputs.
+    pub fn into_transaction(self) -> Transaction {
+        Transaction {
+            version: self.version,
+            input: self.inputs.into_iter().map(|input| input.txin).collect(),
+            output: self.outputs,
+            lock_time: self.lock_time,
+        }
+    }
+
+    /// Creates non-extended transaction from extended transaction.
+    ///
+    /// As opposed to `into_transaction` this allocates for each input and output however, as
+    /// unlike `tx.clone().into_transaction()` this avoids cloning the extra input data.
+    pub fn to_transaction(&self) -> Transaction {
+        Transaction {
+            version: self.version,
+            input: self.inputs.iter().map(|input| input.txin.clone()).collect(),
+            output: self.outputs.clone(),
+            lock_time: self.lock_time,
+        }
+    }
+
+    /// Computes the [`Txid`].
+    ///
+    /// Hashes the transaction **excluding** the segwit data (i.e. the marker, flag bytes, and the
+    /// witness fields themselves). For non-segwit transactions which do not have any segwit data,
+    /// this will be equal to [`Transaction::wtxid()`].
+    pub fn txid(&self) -> Txid {
+        use crate::consensus::Encodable;
+        use crate::hashes::Hash;
+
+        let mut enc = Txid::engine();
+        self.version.consensus_encode(&mut enc).expect("engines don't error");
+        VarInt(self.inputs.len() as u64).consensus_encode(&mut enc).expect("engines don't error");
+        for input in &self.inputs {
+            input.txin.consensus_encode(&mut enc).expect("engines don't error");
+        }
+        self.outputs.consensus_encode(&mut enc).expect("engines don't error");
+        self.lock_time.consensus_encode(&mut enc).expect("engines don't error");
+        Txid::from_engine(enc)
+    }
+
+    /// Returns the fee this transaction is paying.
+    ///
+    /// Returns `None` in case of overflow (creating bitcoins out of thin air).
+    pub fn fee(&self) -> Option<Amount> {
```

Comment:
I'm probably missing something, but could we just have a fee on `TxIn` (extended)? and then the fee for a transaction would be the sum of TxIn fees?  Otherwise this doesn't provide the interface for calculating per txin fee nor long_term_fee.

Reply:
Yes, however during fee estimation it's useful, like computing how wasteful it is to add a UTXO to a new transaction: https://github.com/bitcoin/bitcoin/blob/03752444cd54df05a731557968d5a9f33a55c55c/src/wallet/coinselection.cpp#L165